src/osd/PeeringState.cc

```diff
@@ -4940,6 +4940,7 @@ boost::statechart::result PeeringState::Peering::react(const AdvMap& advmap)
   psdout(10) << "Peering advmap" << dendl;
   if (prior_set.affected_by_map(*(advmap.osdmap), ps->dpp)) {
     psdout(1) << "Peering, affected_by_map, going to Reset" << dendl;
+    ps->clear_primary_state();
```

Comment:
What these two reactions have in common is that they are detecting map changes that *are not interval changes*, so we won't go through start_peering_interval.

For this one, we're detecting prior_set members going down which won't be an interval change unless the member is also an up/acting member.

For the GetLog handler, we're detecting whether the auth_log_shard happened to go down which also wouldn't be an interval change unless the shard happened to be a in the acting/up set.

However, we don't want to have to sprinkle clear_primary_state() calls in each of these handlers -- that's a ready recipe for missing one later.  The right place to add it would be Primary::exit() -- that gets called regardless of how we leave the Primary state (note that both of these transit< Reset >()).

Primary::exit() actually already calls clear_primary_state():

```
void PeeringState::Primary::exit()
{
  context< PeeringMachine >().log_exit(state_name, enter_time);
  DECLARE_LOCALS;
  ps->want_acting.clear();
  utime_t dur = ceph_clock_now() - enter_time;
  pl->get_peering_perf().tinc(rs_primary_latency, dur);
  pl->clear_primary_state();
  ps->state_clear(PG_STATE_CREATING);
}
```

Reply:
Something is missing in that explanation.  osd.9 would have been the primary and as explained above would have invoked clear_primary_state() when it noticed that osd.5 went up (unless I'm missing something?)  That sequence should have no effect on the replicas -- they don't actually rollforward or back until activation (unless I'm missing something).  osd.0 wouldn't have done either until even later since it wasn't in the acting set.